rados/rados_write_op_setxattr.go

```diff
@@ -0,0 +1,58 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package rados
+
+// #cgo LDFLAGS: -lrados
+// #include <rados/librados.h>
+// #include <stdlib.h>
+//
+import "C"
+
+import (
+	"unsafe"
+)
+
+type writeOpSetXattrStep struct {
+	withoutUpdate
+
+	// C arguments
+
+	cName  *C.char
+	cValue *C.char
+}
+
+func newWriteOpSetXattrStep(name string, value []byte) *writeOpSetXattrStep {
+	return &writeOpSetXattrStep{
+		cName:  C.CString(name),
+		cValue: (*C.char)(C.CBytes(value)),
+	}
+}
+
+func (s *writeOpSetXattrStep) free() {
+	C.free(unsafe.Pointer(s.cName))
+	C.free(unsafe.Pointer(s.cValue))
+
+	s.cName = nil
+	s.cValue = nil
+}
+
+// Set an xattr.
+//  PREVIEW
+//
+// Implements:
+//  void rados_write_op_setxattr(rados_write_op_t write_op,
+//                               const char * name,
+//                               const char * value,
+//                               size_t value_len)
+func (w *WriteOp) SetXattr(name string, value []byte) {
+	s := newWriteOpSetXattrStep(name, value)
+	w.steps = append(w.steps, s)
+
+	C.rados_write_op_setxattr(
+		w.op,
+		s.cName,
+		s.cValue,
+		C.size_t(len(value)),
+	)
```

Comment:
Why delay the free of the `cName` and `cValue` until the `free()` call? I was very sure that `rados_write_op_setxattr` would not take ownership of the pointers and instead would create copies for use after the call returns, otherwise I would consider that a bug of librados. Finding the proof was not easy though, but after diving deep through the ceph caves I finally [found it](https://github.com/ceph/ceph/blob/80ae7c307e92763f14f573a8882df41a5c84a21b/src/common/buffer.cc#L595).

Which means: you can just create a `C.CString()`and free it right after the C call. The `value` buffer you don't have to copy at all, just use `&value[0]` as usual.

Reply:
Ah cool and a bit surprising, I didn't know that it's doing a copy, thanks. I'll change it. Is there some general rule of thumb where librados creates a copy instead of just passing a pointer? Is it all these `oid` and `name` (null-terminated) strings? It's not mentioned in librados docs from what I skimmed over.

Reply:
Because it's not mentioned, I assumed it must make a copy. I think that is a general convention in C calls, that if you pass a pointer, and it doesn't take ownership, it will copy everything it might need later. Of course it would be better to always document that explicitly, but well. 🤷